### PR TITLE
#18: Move provisioning to cloud-init

### LINKED ISSUE #18: Move provisioning to cloud-init
Move the docker+docker-compose+cloudwatch-agent installation in user_data (aka cloud-init)

### PR DESCRIPTION
**Pull request by bennesp**
Closes [#2520309](https://buildo.kaiten.io/space/32111/card/2520309)

⚠️ The line above was added during the automatic migration of all github project issues to Kaiten. The old GH issue link is kept here for reference (it is now deprecated, continue follow the original issue on Kaiten).

Closes #18

# Things done

- Moved installation of docker and docker-compose and cloudwatch agent to user_data
  - provisioner `remote-exec` replaced with `init.sh`
  - provisioner `file` replaced with the cat inside `init.sh`
- No more need for `ssh_private_key`, so I removed it everywhere!
- Using terraform 0.14.9, it upgraded things:
  - replaced `.terraform-version` with `versions.tf` and `.terraform.lock.hcl`
  - add quotes to every resource/variable name (sorry for the ugly PR 😞  )

# Test Plan

## tests performed

- Spawned a test instance
- docker is installed
- docker-compose is installed
- cloudwatch agent json file is copied
- cloudwatch agent container is started
- can see metric in cloudwatch

## tests not performed (domain coverage)

Not tested with alarm enabled

### CODE REVIEW

File: cloud-init.tf [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        - name: ubuntu
+          groups: docker
+      write_files:
+        - path: /etc/cwagentconfig/cwagentconfig.json
+          encoding: b64
+          content: ${base64encode(file(coalesce(var.cloudwatch_agent_config, "${path.module}/cwagentconfig.json")))}
+          owner: root:root
+          permissions: '0644'
+      runcmd:
+        - [sudo, docker, run, -d, --restart, always, -v, /etc/cwagentconfig:/etc/cwagentconfig, amazon/cloudwatch-agent:1.247347.6b250880]
```
Comment (mcastellini) [inline cloud-init.tf]: Just 2 things here:
  - why are we using `sudo` here? AFAICT cloud-init runs as root, I would have expected something with `runuser`; 😄 
  - wouldn't it be better to use `amazon/cloudwatch-agent:latest` as a image for the CloudWatch agent?

**Review by mcastellini — CHANGES_REQUESTED** [review]
The PR is OK, just a couple of questions inline and one additional thing: we need to remove references to `ssh_private_key` from the README.

File: cloud-init.tf (hunk 1 above)
Comment (bennesp) [reply cloud-init.tf]: - A sad copy-paste from before, will remove `sudo`
- I maybe saw once that `latest` broke some features in `cloudwatch-agent` image, even if I can't remember which version was. In general using `latest` can be dangerous if the image has breaking changes. Anyway I don't have strong opinions on using or non using latest, so if you suggest to use it, I can freely change it

**Comment by bennesp** [comment]
Moreover, references to `ssh_private_key` in the README are already removed with this PR, aren't they?

**Comment by mcastellini** [comment]
> Moreover, references to `ssh_private_key` in the README are already removed with this PR, aren't they?

Hi, I am blind apparently.

File: cloud-init.tf (hunk 1 above)
Comment (mcastellini) [reply cloud-init.tf]: Let's be conservative and stick to a version then, the image does not seem to provide a decent minor version tag.

**Review by mcastellini — APPROVED** [review]

**Pull request merged**